### PR TITLE
Sort results on replica, merge on envd

### PR DESCRIPTION
**Pull request by antiguru**
Sort results on replica, merge on environmentd.

Previously, we'd sort data only on evironmentd, which would cause it to consume more CPU than necessary. This change moves some of the sorting to clusterd, and only leaves the last merge step on environmentd.

The PR selects a minimal approach, and leaves most of the code related to result finishing untouched. It introduces an invariant that peek results must always be sorted according to the finishing, anything else will lead to undefined results. However, there's nothing that enforces the results to be sorted with the same ordering, which is potentially bad. Inside environmentd, it uses a simple heap to combine $k$ sorted runs into a single permutation map.

The interfaces to `RowCollection` (`new`, `sorted_view`) now take a `&[ColumnOrder]`, and internally the implementation picks the right comparison function. If the column order slice is empty, it'll skip decoding the rows and directly defer to the tiebreaker.

The PR moves the `RowCollection` type into `mz-expr`, which isn't ideal. This is required because the `ColumnOrder` type is defined here, and we'd like to pass it to the constructor of the type. Alternatives would be to have a function here that passes the correct comparison function to `RowCollection`, but that seems to be strictly worse than moving the type.
I considered moving the type to `compute-types`, which seems a better fit, but not all uses of `RowCollection` depend on `compute-types`. If this is upsetting, I can think about alternatives.

This complexity for sorting on the cluster is roughly $\frac{n}{k}\cdot\log \frac{n}{k}$, where $n$ is the total number of result records, and $k$ the number of workers. The last merge step then has a time complexity of $n\cdot\log k$ to combine $k$ sorted runs into one.

Follow-up items include:
* Avoid a single `Bytes` allocation for all rows, and instead keep the individual allocations.
* Assert that all `RowCollections` are sorted equally.
* Move the binary heap into an iterator to avoid materializing the sorted view permutation.

### Tips to the reviewer

Don't look at individual commits.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/repr/src/row/collection.rs [hunk 1]
```diff
@@ -132,16 +142,48 @@ impl RowCollection {
...
+
+        let mut view = Vec::with_capacity(self.metadata.len());
+
+        while let Some(Reverse(mut finger)) = heap.pop() {
+            view.push(finger.start);
+            finger.start += 1;
+            if finger.start < finger.end {
+                heap.push(Reverse(finger));
+            }
+        }
```
Comment (ParkMyCar) [inline src/repr/src/row/collection.rs:184]: It feels like there is a great opportunity to push this logic into `SortedRowCollection` or `SortedRowCollectionIter` maybe? i.e. as folks iterate through a row collection is when we do this streaming merge sort?

File: src/repr/src/row/collection.proto [hunk 2]
```diff
@@ -14,6 +14,7 @@ package mz_repr.row.collection;
 message ProtoRowCollection {
   bytes encoded = 1;
   repeated ProtoEncodedRowMetadata metadata = 2;
+  repeated uint64 fingers = 3;
```
Comment (ParkMyCar) [inline src/repr/src/row/collection.proto:17]: "fingers" is an interesting term for this, what do you think about calling this field "runs"? It matches terminology from Persist for a similar concept

File: src/repr/src/row/collection.rs [hunk 3]
```diff
@@ -33,6 +35,8 @@ pub struct RowCollection {
     encoded: Bytes,
     /// Metadata about an individual Row in the blob.
     metadata: Vec<EncodedRowMetadata>,
+    /// Start of sorted runs of rows in rows.
+    fingers: Vec<usize>,
```
Comment (teskje) [inline src/repr/src/row/collection.rs:39]: The documentation here confused me. This field actually stores the indexes of the _ends_ of sorted runs, right? Is there a reason for that? It does feel like storing the start indexes would be more natural.

I know you said the PR lacks documentation, so if you still planned to adjust it here then nvm!

File: src/repr/src/row/collection.rs (hunk 1 above)
Comment (teskje) [reply src/repr/src/row/collection.rs:184]: Agreed! I assume we're going to iterate through the result rows only once when we send them over the wire, so we can avoid having the extra `view` buffer around and decoding the rows twice.

**Comment by teskje** [comment]
We discussed a bit offline already, an my understanding is that this PR is meant to stop the bleeding with an as-small-as-possible diff and do improvements as a follow-up. That plan is fine with me.

> IMO the biggest feedback I have is enforcing the invariant that peek results must sorted at the type level

I agree with that. Once concern is that we add some place where we create `PeekResponse::Rows` but forget that the contained rows are expected to be sorted runs. Currently there is an assert in the merging but we don't want to keep that in prod, so it'd be easy to end up with incorrect results returned to the user. So `PeekResponse::Rows` should contain a type that ensures the sorting invariants we need.

Another reason for wanting such a type is that we sometimes don't have to sort! Specifically, if the `order_by` is empty, we can return the data in any order, I _think_. I initially thought that wasn't the case because results from different workers can cancel out, but since their diffs are `NonZeroU64`, they can only add up but never cancel. Don't trust me on this, I'm probably missing something about how peek finishing works.

But _if_ it's true that we don't need to sort if the `order_by` is empty, then we want a type that knows about the `order_by` and does the right thing (sorting or not) depending on it. For example:

```rust
struct RowRuns {
    runs: Vec<RowCollection>,
    order_by: Vec<ColumnOrder>,
}

impl RowRuns {
    fn push(&mut self, mut rows: Vec<(Row, NonZeroU64)>) {
        if !self.order_by.is_empty() {
            sort(&mut rows, &self.order_by);
        }
        self.runs.push(RowCollection::new(&rows));
    }
}
```

**Comment by antiguru** [comment]
> Another reason for wanting such a type is that we sometimes don't have to sort!

I agree that strictly speaking there are cases where we don't have to sort, but I'm not comfortable changing the invariant as part of this PR. We might have downstream code that relies on a certain row order, as well as our tests, so I'd like to separate this from the current effort.

File: src/repr/src/row/collection.rs (hunk 1 above)
Comment (antiguru) [reply src/repr/src/row/collection.rs]: Unfortunately, we need to iterate multiple times to determine the size of the whole result. I agree we shouldn't (and there is no deep reason we have to), but it requires more changes.

File: src/repr/src/row/collection.proto (hunk 2 above)
Comment (antiguru) [reply src/repr/src/row/collection.proto:17]: Changing to `runs`.

**Comment by antiguru** [comment]
> It kind of feels like RowCollection could naturally become a SortedRowRun and then SortedRowCollection becomes a collection of SortedRowRuns?

I agree it could! It's a non-trivial departure to what we currently have: At the moment, we allow to index into a `RowCollection` and the sorted variants, which we use primarily to iterate. If we want to avoid the index lookup, we could change the iterator to sit on the binary heap, but then we'd need to be careful not to clone the iterator -- the cost of iterating would be $n\log k$ instead of $n$.

**Comment by teskje** [comment]
> I agree that strictly speaking there are cases where we don't have to sort, but I'm not comfortable changing the invariant as part of this PR. We might have downstream code that relies on a certain row order, as well as our tests, so I'd like to separate this from the current effort.

Yes, I'm very much in favor of taking small steps! Just wanted to record my thoughts for follow-ups we can/should do. Also partly to check my thinking around whether or not sorting is necessary.

**Comment by antiguru** [comment]
Nightly run: https://buildkite.com/materialize/nightly/builds/10466

**Review by teskje — APPROVED** [review]
LGTM!

**Review by ParkMyCar — APPROVED** [review]
LGTM!

Sorry I didn't realize pushing the sort into `RowCollection` would require moving the struct to the `expr` crate 🙈 thanks for making that change!

**Comment by antiguru** [comment]
Thanks for the reviews!

**Pull request merged**